[Safer CPP] Address issues in ColorSpaceCG

ruthvikkonda opened the pull request:
#### ec77ec4587d561c161d193b086dbc46fbc8b5b24
<pre>
[Safer CPP] Address issues in ColorSpaceCG
<a href="https://bugs.webkit.org/show_bug.cgi?id=293643">https://bugs.webkit.org/show_bug.cgi?id=293643</a>
<a href="https://rdar.apple.com/152111464">rdar://152111464</a>

Reviewed by Chris Dumez.

Address SaferCPP issues in ColorSpaceCG. Refactor files to support function name changes in ColorSpaceCG. Fix SaferCPP issues in a few other files via refactoring in ColorSpaceCG.

* Source/WebCore/Modules/system-preview/ARKitBadgeSystemImage.mm:
(WebCore::ARKitBadgeSystemImage::draw const):
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::createColor):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::SRGB):
(WebCore::DestinationColorSpace::LinearSRGB):
(WebCore::DestinationColorSpace::DisplayP3):
(WebCore::DestinationColorSpace::ExtendedDisplayP3):
(WebCore::DestinationColorSpace::ExtendedSRGB):
(WebCore::DestinationColorSpace::ExtendedRec2020):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createImageForTimeInRect):
* Source/WebCore/platform/graphics/ca/cocoa/WebSystemBackdropLayer.mm:
(-[WebLightSystemBackdropLayer init]):
(-[WebDarkSystemBackdropLayer init]):
* Source/WebCore/platform/graphics/cg/ColorSpaceCG.cpp:
(WebCore::namedColorSpace):
(WebCore::extendedNamedColorSpace):
(WebCore::sRGBColorSpaceSingleton):
(WebCore::adobeRGB1998ColorSpaceSingleton):
(WebCore::displayP3ColorSpaceSingleton):
(WebCore::extendedAdobeRGB1998ColorSpaceSingleton):
(WebCore::extendedDisplayP3ColorSpaceSingleton):
(WebCore::extendedITUR_2020ColorSpaceSingleton):
(WebCore::extendedLinearSRGBColorSpaceSingleton):
(WebCore::extendedROMMRGBColorSpaceSingleton):
(WebCore::extendedSRGBColorSpaceSingleton):
(WebCore::ITUR_2020ColorSpaceSingleton):
(WebCore::linearSRGBColorSpaceSingleton):
(WebCore::ROMMRGBColorSpaceSingleton):
(WebCore::xyzD50ColorSpaceSingleton):
(WebCore::colorSpaceForCGColorSpace):
(WebCore::sRGBColorSpaceRef): Deleted.
(WebCore::adobeRGB1998ColorSpaceRef): Deleted.
(WebCore::displayP3ColorSpaceRef): Deleted.
(WebCore::extendedAdobeRGB1998ColorSpaceRef): Deleted.
(WebCore::extendedDisplayP3ColorSpaceRef): Deleted.
(WebCore::extendedITUR_2020ColorSpaceRef): Deleted.
(WebCore::extendedLinearSRGBColorSpaceRef): Deleted.
(WebCore::extendedROMMRGBColorSpaceRef): Deleted.
(WebCore::extendedSRGBColorSpaceRef): Deleted.
(WebCore::ITUR_2020ColorSpaceRef): Deleted.
(WebCore::linearSRGBColorSpaceRef): Deleted.
(WebCore::ROMMRGBColorSpaceRef): Deleted.
(WebCore::xyzD50ColorSpaceRef): Deleted.
* Source/WebCore/platform/graphics/cg/ColorSpaceCG.h:
(WebCore::CGColorSpaceMapping&lt;ColorSpace::SRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::A98RGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::DisplayP3&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedA98RGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedDisplayP3&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedRec2020&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedLinearSRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedProPhotoRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ExtendedSRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::Rec2020&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::LinearSRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::ProPhotoRGB&gt;::colorSpace):
(WebCore::CGColorSpaceMapping&lt;ColorSpace::XYZ_D50&gt;::colorSpace):
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createCGColorSpaceForCVPixelBuffer):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::createPixelBuffer):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::drawMenuListBackground):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.cpp:
(WebCore::sRGBColorSpaceSingleton):
(WebCore::linearSRGBColorSpaceSingleton):
(WebCore::displayP3ColorSpaceSingleton):
(WebCore::sRGBColorSpaceRef): Deleted.
(WebCore::linearSRGBColorSpaceRef): Deleted.
(WebCore::displayP3ColorSpaceRef): Deleted.
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.h:
* Source/WebCore/platform/ios/ColorIOS.mm:
(WebCore::colorFromCocoaColor):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::initializeDebugIndicator):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(imageFromRect):

Canonical link: <a href="https://commits.webkit.org/295849@main">https://commits.webkit.org/295849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f83f3b13e70fc0e0af78d02d1468c91bb15c97dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106123 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111320 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108162 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34375 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80609 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95767 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60947 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20532 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13870 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56158 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90325 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13905 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24540 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89689 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33625 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92000 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89381 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22838 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34246 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12061 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29026 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33186 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38598 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32932 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36282 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34530 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/graphics/cg/ColorSpaceCG.cpp:
```diff
@@ -40,7 +40,7 @@ template<const CFStringRef& colorSpaceNameGlobalConstant> static CGColorSpaceRef
     static std::once_flag onceFlag;
     std::call_once(onceFlag, [] {
-        colorSpace.get() = adoptCF(CGColorSpaceCreateWithName(colorSpaceNameGlobalConstant));
+        colorSpace.get() = adoptCF(CGColorSpaceCreateWithName(RetainPtr { colorSpaceNameGlobalConstant }.get()));
```
`NeverDestroyed` is designed for immediate initialization. Initializing to null and then assigning later is bad code smell. To avoid the `.get() =` (which we maybe need to prohibit eventually), use `LazyNeverDestroyed` and call `construct`.

geoffreygaren commented on Source/WebCore/platform/graphics/cg/ColorSpaceCG.cpp:
```diff
@@ -52,7 +52,7 @@ template<const CFStringRef& colorSpaceNameGlobalConstant> static CGColorSpaceRef
     static std::once_flag onceFlag;
     std::call_once(onceFlag, [] {
-        colorSpace.get() = adoptCF(CGColorSpaceCreateExtended(namedColorSpace<colorSpaceNameGlobalConstant>()));
+        colorSpace.get() = adoptCF(CGColorSpaceCreateExtended(RetainPtr { namedColorSpace<colorSpaceNameGlobalConstant>() }.get()));
```
Use `LazyNeverDestroyed` and `construct`.

geoffreygaren commented on Source/WebCore/platform/graphics/cg/ColorSpaceCG.cpp:
```diff
@@ -157,62 +157,62 @@ std::optional<ColorSpace> colorSpaceForCGColorSpace(CGColorSpaceRef colorSpace)
     // FIXME: Consider using a UncheckedKeyHashMap (with CFHash based keys) rather than the linear set of tests.
 
-    if (CGColorSpaceEqualToColorSpace(colorSpace, sRGBColorSpaceRef()))
+    if (CGColorSpaceEqualToColorSpace(colorSpace, RetainPtr { sRGBColorSpaceRef() }.get()))
```
Better to rename `sRGBColorSpaceRef` to `sRGBColorSpaceSingleton`. Then static analysis will acknowledge that you're returning a global, and you won't need the `RetainPtr`.

geoffreygaren replied:
Same comment in many other places.

geoffreygaren reviewed (changes requested):
I think you've met all our Safer CPP guidelines here, but I want to suggest some subtle ways to improve code quality.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/06db015add9affe06bd759cad0856abadc7a3787. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45964

ruthvikkonda commented on Source/WTF/wtf/NeverDestroyed.h:
```diff
@@ -124,6 +124,10 @@ template<typename T, typename AccessTraits> class LazyNeverDestroyed {
     template<typename... Args>
     void construct(Args&&... args)
     {
+        ASSERT(!m_isConstructed);
```
Per Geoff's recommendation, added this to mechanically enforce against double-initialization

cdumez commented:
EWS is all red so I'm holding off on reviewing.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e7db6d728f91907af1387a28f4264741d093ff88. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45964

ruthvikkonda commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -2034,7 +2034,7 @@ static URL conformFragmentIdentifierForURL(const URL& url)
 ALLOW_DEPRECATED_DECLARATIONS_BEGIN
     RetainPtr<CGImageRef> rawImage = adoptCF([m_imageGenerator copyCGImageAtTime:PAL::CMTimeMakeWithSeconds(time, 600) actualTime:nil error:nil]);
-    RetainPtr<CGImageRef> image = adoptCF(CGImageCreateCopyWithColorSpace(rawImage.get(), sRGBColorSpaceRef()));
+    RetainPtr<CGImageRef> image = adoptCF(CGImageCreateCopyWithColorSpace(rawImage.get(), sRGBColorSpaceSingleton()));
```
Fixed!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/82cf0683ccb23bc0be54a2d279326c9ff20212c1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45964

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f83f3b13e70fc0e0af78d02d1468c91bb15c97dd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45964

ruthvikkonda commented:
Addressed all new comments.

ruthvikkonda committed:
[Safer CPP] Address issues in ColorSpaceCG
https://bugs.webkit.org/show_bug.cgi?id=293643
rdar://152111464

Reviewed by Chris Dumez.

Address SaferCPP issues in ColorSpaceCG. Refactor files to support function name changes in ColorSpaceCG. Fix SaferCPP issues in a few other files via refactoring in ColorSpaceCG.

* Source/WebCore/Modules/system-preview/ARKitBadgeSystemImage.mm:
(WebCore::ARKitBadgeSystemImage::draw const):
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::createColor):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::SRGB):
(WebCore::DestinationColorSpace::LinearSRGB):
(WebCore::DestinationColorSpace::DisplayP3):
(WebCore::DestinationColorSpace::ExtendedDisplayP3):
(WebCore::DestinationColorSpace::ExtendedSRGB):
(WebCore::DestinationColorSpace::ExtendedRec2020):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createImageForTimeInRect):
* Source/WebCore/platform/graphics/ca/cocoa/WebSystemBackdropLayer.mm:
(-[WebLightSystemBackdropLayer init]):
(-[WebDarkSystemBackdropLayer init]):
* Source/WebCore/platform/graphics/cg/ColorSpaceCG.cpp:
(WebCore::namedColorSpace):
(WebCore::extendedNamedColorSpace):
(WebCore::sRGBColorSpaceSingleton):
(WebCore::adobeRGB1998ColorSpaceSingleton):
(WebCore::displayP3ColorSpaceSingleton):
(WebCore::extendedAdobeRGB1998ColorSpaceSingleton):
(WebCore::extendedDisplayP3ColorSpaceSingleton):
(WebCore::extendedITUR_2020ColorSpaceSingleton):
(WebCore::extendedLinearSRGBColorSpaceSingleton):
(WebCore::extendedROMMRGBColorSpaceSingleton):
(WebCore::extendedSRGBColorSpaceSingleton):
(WebCore::ITUR_2020ColorSpaceSingleton):
(WebCore::linearSRGBColorSpaceSingleton):
(WebCore::ROMMRGBColorSpaceSingleton):
(WebCore::xyzD50ColorSpaceSingleton):
(WebCore::colorSpaceForCGColorSpace):
(WebCore::sRGBColorSpaceRef): Deleted.
(WebCore::adobeRGB1998ColorSpaceRef): Deleted.
(WebCore::displayP3ColorSpaceRef): Deleted.
(WebCore::extendedAdobeRGB1998ColorSpaceRef): Deleted.
(WebCore::extendedDisplayP3ColorSpaceRef): Deleted.
(WebCore::extendedITUR_2020ColorSpaceRef): Deleted.
(WebCore::extendedLinearSRGBColorSpaceRef): Deleted.
(WebCore::extendedROMMRGBColorSpaceRef): Deleted.
(WebCore::extendedSRGBColorSpaceRef): Deleted.
(WebCore::ITUR_2020ColorSpaceRef): Deleted.
(WebCore::linearSRGBColorSpaceRef): Deleted.
(WebCore::ROMMRGBColorSpaceRef): Deleted.
(WebCore::xyzD50ColorSpaceRef): Deleted.
* Source/WebCore/platform/graphics/cg/ColorSpaceCG.h:
(WebCore::CGColorSpaceMapping<ColorSpace::SRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::A98RGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::DisplayP3>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedA98RGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedDisplayP3>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedRec2020>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedLinearSRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedProPhotoRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ExtendedSRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::Rec2020>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::LinearSRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::ProPhotoRGB>::colorSpace):
(WebCore::CGColorSpaceMapping<ColorSpace::XYZ_D50>::colorSpace):
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createCGColorSpaceForCVPixelBuffer):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::createPixelBuffer):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::drawMenuListBackground):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.cpp:
(WebCore::sRGBColorSpaceSingleton):
(WebCore::linearSRGBColorSpaceSingleton):
(WebCore::displayP3ColorSpaceSingleton):
(WebCore::sRGBColorSpaceRef): Deleted.
(WebCore::linearSRGBColorSpaceRef): Deleted.
(WebCore::displayP3ColorSpaceRef): Deleted.
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.h:
* Source/WebCore/platform/ios/ColorIOS.mm:
(WebCore::colorFromCocoaColor):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::initializeDebugIndicator):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(imageFromRect):

Canonical link: https://commits.webkit.org/295849@main
